File: src/rule-engine.ts
import {
  ResponseValueSingleType,
  ResponseValueType
} from '@iqbspecs/response/response.interface';
import {
  CodingRule,
  ProcessingParameterType,
  RuleSet,
  TransformedResponseValueType
} from '@iqbspecs/coding-scheme/coding-scheme.interface';
import {
  getValueAsNumber,
  isEmptyValue,
  transformString
} from './value-transform';

function findString(
  value: string,
  codingProcessing: ProcessingParameterType[],
  parameters: string[] = []
): boolean {
  const allStrings = parameters
    .flatMap(p => p.split(/\r?\n/))
    .map(s => {
      const transformed = transformString(s, codingProcessing);
      return Array.isArray(transformed) ? transformed[0] || '' : transformed;
    });

  let stringToCompare = transformString(value, codingProcessing);
  if (Array.isArray(stringToCompare)) {
    stringToCompare = stringToCompare[0] || '';
  }

  return allStrings.includes(stringToCompare);
}

function findStringRegEx(
  value: string,
  parameters: string[],
  addCaseIgnoreFlag: boolean
): boolean {
  const allStrings = parameters.flatMap(p => p.split(/\r?\n/));
  return allStrings.some(s => {
    try {
      const regEx = new RegExp(s, addCaseIgnoreFlag ? 'i' : undefined);
      return regEx.test(value);
    } catch (e) {
      return false;
    }
  });
}

function findNumericValue(
  value: ResponseValueSingleType,
  parameters: string[] = []
): boolean {
  const allCompareValues = parameters
    .flatMap(p => p.split(/\r?\n/))
    .map(s => getValueAsNumber(s));
  const valueAsNumber = getValueAsNumber(value);
  return valueAsNumber !== null && allCompareValues.includes(valueAsNumber);
}

export function checkOneValue(
  valueToCheck: ResponseValueSingleType,
  rule: CodingRule,
  codingProcessing: ProcessingParameterType[]
): boolean {
  let returnValue = false;
  let valueAsNumber: number | null = null;

  // eslint-disable-next-line default-case
  switch (rule.method) {
    case 'IS_NULL':
      if (valueToCheck === null) returnValue = true;
      break;
    case 'IS_EMPTY':
      if (isEmptyValue(valueToCheck as ResponseValueType)) returnValue = true;
      break;
    case 'MATCH':
      if (valueToCheck !== null && valueToCheck !== '') {
        let valueAsString = valueToCheck;
        if (typeof valueToCheck === 'number') {
          valueAsString = valueToCheck.toString(10);
        } else if (typeof valueToCheck === 'boolean') {
          valueAsString = valueToCheck.toString();
        }
        returnValue = findString(
          valueAsString as string,
          codingProcessing,
          rule.parameters
        );
      }
      break;
    case 'MATCH_REGEX':
      if (valueToCheck !== null && valueToCheck !== '') {
        let valueAsString = valueToCheck;
        if (typeof valueToCheck === 'number') {
          valueAsString = valueToCheck.toString(10);
        } else if (typeof valueToCheck === 'boolean') {
          valueAsString = valueToCheck.toString();
        }
        returnValue = findStringRegEx(
          valueAsString as string,
          rule.parameters || [],
          codingProcessing.includes('IGNORE_CASE')
        );
      }
      break;
    case 'NUMERIC_MATCH':
      returnValue = findNumericValue(valueToCheck, rule.parameters);
      break;
    case 'NUMERIC_LESS_THAN':
      if (valueToCheck !== null && valueToCheck !== '') {
        valueAsNumber = getValueAsNumber(valueToCheck);
        if (typeof valueAsNumber === 'number' && rule.parameters) {
          const compareValue = Number.parseFloat(rule.parameters[0]);
          returnValue =
            !Number.isNaN(compareValue) && valueAsNumber < compareValue;
        }
      }
      break;
    case 'NUMERIC_MAX':
      if (valueToCheck !== null && valueToCheck !== '') {
        valueAsNumber = getValueAsNumber(valueToCheck);
        if (typeof valueAsNumber === 'number' && rule.parameters) {
          const compareValue = Number.parseFloat(rule.parameters[0]);
          returnValue =
            !Number.isNaN(compareValue) && valueAsNumber <= compareValue;
        }
      }
      break;
    case 'NUMERIC_MORE_THAN':
      if (valueToCheck !== null && valueToCheck !== '') {
        valueAsNumber = getValueAsNumber(valueToCheck);
        if (typeof valueAsNumber === 'number' && rule.parameters) {
          const compareValue = Number.parseFloat(rule.parameters[0]);
          returnValue =
            !Number.isNaN(compareValue) && valueAsNumber > compareValue;
        }
      }
      break;
    case 'NUMERIC_MIN':
      if (valueToCheck !== null && valueToCheck !== '') {
        valueAsNumber = getValueAsNumber(valueToCheck);
        if (typeof valueAsNumber === 'number' && rule.parameters) {
          const compareValue = Number.parseFloat(rule.parameters[0]);
          returnValue =
            !Number.isNaN(compareValue) && valueAsNumber >= compareValue;
        }
      }
      break;
    case 'NUMERIC_RANGE':
      if (valueToCheck !== null && valueToCheck !== '') {
        valueAsNumber = getValueAsNumber(valueToCheck);
        if (typeof valueAsNumber === 'number' && rule.parameters) {
          const compareValueLL = getValueAsNumber(rule.parameters[0]);
          const compareValueUL = getValueAsNumber(rule.parameters[1]);
          if (
            typeof compareValueLL === 'number' &&
            typeof compareValueUL === 'number'
          ) {
            returnValue =
              valueAsNumber > compareValueLL && valueAsNumber <= compareValueUL;
          }
        }
      }
      break;
    case 'NUMERIC_FULL_RANGE':
      if (valueToCheck !== null && valueToCheck !== '') {
        valueAsNumber = getValueAsNumber(valueToCheck);
        if (typeof valueAsNumber === 'number' && rule.parameters) {
          const compareValueLL = getValueAsNumber(rule.parameters[0]);
          const compareValueUL = getValueAsNumber(rule.parameters[1]);
          if (
            typeof compareValueLL === 'number' &&
            typeof compareValueUL === 'number'
          ) {
            returnValue =
              valueAsNumber >= compareValueLL &&
              valueAsNumber <= compareValueUL;
          }
        }
      }
      break;
    case 'IS_TRUE':
      returnValue =
        valueToCheck === 1 ||
        valueToCheck === '1' ||
        valueToCheck === true ||
        valueToCheck === 'true';
      break;
    case 'IS_FALSE':
      returnValue =
        valueToCheck === 0 ||
        valueToCheck === '0' ||
        valueToCheck === false ||
        valueToCheck === 'false';
      break;
  }

  return returnValue;
}

function isMatchRule(
  valueToCheck: TransformedResponseValueType,
  rule: CodingRule,
  isValueArray: boolean,
  codingProcessing: ProcessingParameterType[]
): boolean {
  if (Array.isArray(valueToCheck) && isValueArray) {
    if (valueToCheck.length === 0) {
      return checkOneValue('', rule, codingProcessing);
    }
    return valueToCheck.some(valueMember => {
      if (Array.isArray(valueMember)) {
        if (typeof rule.fragment === 'undefined' || rule.fragment < 0) {
          return valueMember.some(fragment => checkOneValue(
            fragment as ResponseValueSingleType,
            rule,
            codingProcessing
          )
          );
        }

        return checkOneValue(
          valueMember[rule.fragment] as ResponseValueSingleType,
          rule,
          codingProcessing
        );
      }

      return checkOneValue(
        valueMember as ResponseValueSingleType,
        rule,
        codingProcessing
      );
    });
  }

  if (Array.isArray(valueToCheck)) {
    const { fragment } = rule;
    if (fragment == null || fragment < 0) {
      return valueToCheck.some(value => checkOneValue(value as ResponseValueSingleType, rule, codingProcessing)
      );
    }

    return checkOneValue(
      valueToCheck[fragment] as ResponseValueSingleType,
      rule,
      codingProcessing
    );
  }

  return checkOneValue(
    valueToCheck as ResponseValueSingleType,
    rule,
    codingProcessing
  );
}

function resolveValueArrayPosMember(
  valueToCheck: TransformedResponseValueType,
  ruleSet: RuleSet,
  isValueArray: boolean
): TransformedResponseValueType | undefined {
  if (!isValueArray || !Array.isArray(valueToCheck)) {
    return undefined;
  }

  if (typeof ruleSet.valueArrayPos === 'number') {
    if (
      ruleSet.valueArrayPos >= 0 &&
      ruleSet.valueArrayPos < valueToCheck.length
    ) {
      return valueToCheck[ruleSet.valueArrayPos];
    }
    return undefined;
  }

  if (ruleSet.valueArrayPos === 'SUM') {
    return valueToCheck
      .map(v => {
        if (Array.isArray(v)) {
          return v
            .map(s => getValueAsNumber(s as ResponseValueSingleType) || 0)
            .reduce((a, b) => a + b, 0);
        }
        return getValueAsNumber(v as ResponseValueSingleType) || 0;
      })
      .reduce((pv, cv) => pv + cv, 0);
  }

  if (ruleSet.valueArrayPos === 'LENGTH') {
    return valueToCheck.length;
  }

  return undefined;
}

function evaluateRuleSetRules(
  valueToEvaluate: TransformedResponseValueType,
  ruleSet: RuleSet,
  isValueArray: boolean,
  codingProcessing: ProcessingParameterType[]
): boolean {
  if (!ruleSet.ruleOperatorAnd) {
    return (ruleSet.rules ?? []).some(rule => isMatchRule(valueToEvaluate, rule, isValueArray, codingProcessing)
    );
  }

  return (ruleSet.rules ?? []).every(rule => isMatchRule(valueToEvaluate, rule, isValueArray, codingProcessing)
  );
}

function evaluateAnyOpen(
  valueToCheck: TransformedResponseValueType,
  ruleSet: RuleSet,
  codingProcessing: ProcessingParameterType[]
): boolean {
  if (!Array.isArray(valueToCheck)) return false;
  if (valueToCheck.length === 0) return false;
  if (ruleSet.valueArrayPos !== 'ANY_OPEN') return false;

  return valueToCheck.some(value => (ruleSet.rules ?? []).every(rule => isMatchRule(value, rule, false, codingProcessing)
  )
  );
}

function evaluateAny(
  valueToCheck: TransformedResponseValueType,
  ruleSet: RuleSet,
  codingProcessing: ProcessingParameterType[]
): boolean {
  if (!Array.isArray(valueToCheck)) return false;
  if (valueToCheck.length <= 1) return false;
  if (ruleSet.valueArrayPos !== 'ANY') return false;

  return valueToCheck.every(value => (ruleSet.rules ?? []).every(rule => isMatchRule(value, rule, false, codingProcessing)
  )
  );
}

export function isMatchRuleSet(
  valueToCheck: TransformedResponseValueType,
  ruleSet: RuleSet,
  isValueArray: boolean,
  codingProcessing: ProcessingParameterType[]
): boolean {
  if (!ruleSet.rules || ruleSet.rules.length === 0) {
    return false;
  }

  const valueMemberToCheck = resolveValueArrayPosMember(
    valueToCheck,
    ruleSet,
    isValueArray
  );

  const valueToEvaluate =
    typeof valueMemberToCheck !== 'undefined' ?
      valueMemberToCheck :
      valueToCheck;

  const isValueArrayForRules =
    typeof valueMemberToCheck === 'undefined' && isValueArray;
  const rulesMatch = evaluateRuleSetRules(
    valueToEvaluate,
    ruleSet,
    isValueArrayForRules,
    codingProcessing
  );

  if (rulesMatch && isValueArray && Array.isArray(valueToCheck)) {
    if (ruleSet.valueArrayPos === 'ANY_OPEN' && valueToCheck.length > 0) {
      return evaluateAnyOpen(valueToCheck, ruleSet, codingProcessing);
    }
    if (ruleSet.valueArrayPos === 'ANY' && valueToCheck.length > 1) {
      return evaluateAny(valueToCheck, ruleSet, codingProcessing);
    }
  }

  return rulesMatch;
}

File: dotnet/src/Iqb.Responses/RuleEngine.cs
using System.Text.RegularExpressions;

namespace Iqb.Responses;

internal static class RuleEngine
{
    private static readonly TimeSpan RegexTimeout = TimeSpan.FromMilliseconds(500);
    private static readonly object JavaScriptUndefined = new();

    public static bool IsMatchRuleSet(
        object? valueToCheck,
        RuleSet ruleSet,
        bool isValueArray,
        IReadOnlyCollection<string> codingProcessing)
    {
        if (ruleSet.Rules.Count == 0)
        {
            return false;
        }

        var member = ResolveArrayMember(valueToCheck, ruleSet, isValueArray, out var hasMember);
        var valueToEvaluate = hasMember ? member : valueToCheck;
        var arrayForRules = !hasMember && isValueArray;
        var matches = ruleSet.RuleOperatorAnd == true
            ? ruleSet.Rules.All(rule => IsMatchRule(valueToEvaluate, rule, arrayForRules, codingProcessing))
            : ruleSet.Rules.Any(rule => IsMatchRule(valueToEvaluate, rule, arrayForRules, codingProcessing));

        if (!matches || !isValueArray || valueToCheck is not IEnumerable<object?> sequence)
        {
            return matches;
        }

        var values = sequence.ToList();
        var position = ruleSet.ValueArrayPos as string;
        if (position == "ANY_OPEN" && values.Count > 0)
        {
            return values.Any(value => ruleSet.Rules.All(rule => IsMatchRule(value, rule, false, codingProcessing)));
        }
        if (position == "ANY" && values.Count > 1)
        {
            return values.All(value => ruleSet.Rules.All(rule => IsMatchRule(value, rule, false, codingProcessing)));
        }
        return matches;
    }

    private static object? ResolveArrayMember(object? value, RuleSet ruleSet, bool isArray, out bool found)
    {
        found = false;
        if (!isArray || value is not IEnumerable<object?> sequence)
        {
            return null;
        }
        var values = sequence.ToList();
        if (ruleSet.ValueArrayPos is double numericPosition)
        {
            var index = (int)numericPosition;
            if (numericPosition == index && index >= 0 && index < values.Count)
            {
                found = true;
                return values[index];
            }
            return null;
        }
        if (ruleSet.ValueArrayPos is not string position)
        {
            return null;
        }
        if (position == "SUM")
        {
            found = true;
            return values.Sum(value => value is IEnumerable<object?> nested
                ? nested.Sum(item => ValueTransforms.GetValueAsNumber(item) ?? 0d)
                : ValueTransforms.GetValueAsNumber(value) ?? 0d);
        }
        if (position == "LENGTH")
        {
            found = true;
            return (double)values.Count;
        }
        return null;
    }

    private static bool IsMatchRule(
        object? value,
        CodingRule rule,
        bool isValueArray,
        IReadOnlyCollection<string> processing)
    {
        if (value is IEnumerable<object?> sequence)
        {
            var values = sequence.ToList();
            if (isValueArray)
            {
                if (values.Count == 0)
                {
                    return CheckOneValue(string.Empty, rule, processing);
                }
                return values.Any(member => MatchArrayMember(member, rule, processing));
            }

            if (rule.Fragment is null or < 0)
            {
                return values.Any(member => CheckOneValue(member, rule, processing));
            }
            var fragmentIndex = (int)rule.Fragment.Value;
            var fragment = rule.Fragment.Value == fragmentIndex && fragmentIndex >= 0 && fragmentIndex < values.Count
                ? values[fragmentIndex]
                : JavaScriptUndefined;
            return CheckOneValue(fragment, rule, processing);
        }
        return CheckOneValue(value, rule, processing);
    }

    private static bool MatchArrayMember(object? member, CodingRule rule, IReadOnlyCollection<string> processing)
    {
        if (member is not IEnumerable<object?> nested)
        {
            return CheckOneValue(member, rule, processing);
        }
        var fragments = nested.ToList();
        if (rule.Fragment is null or < 0)
        {
            return fragments.Any(fragment => CheckOneValue(fragment, rule, processing));
        }
        var fragmentIndex = (int)rule.Fragment.Value;
        var fragment = rule.Fragment.Value == fragmentIndex && fragmentIndex >= 0 && fragmentIndex < fragments.Count
            ? fragments[fragmentIndex]
            : JavaScriptUndefined;
        return CheckOneValue(fragment, rule, processing);
    }

    private static bool CheckOneValue(object? value, CodingRule rule, IReadOnlyCollection<string> processing)
    {
        var parameters = rule.Parameters ?? [];
        if (ReferenceEquals(value, JavaScriptUndefined))
        {
            if (rule.Method.StartsWith("NUMERIC_", StringComparison.Ordinal))
                throw new InvalidOperationException("A numeric rule cannot evaluate an undefined fragment.");
            return rule.Method == "MATCH_REGEX" && FindRegex("undefined", parameters, processing.Contains("IGNORE_CASE"));
        }
        switch (rule.Method)
        {
            case "IS_NULL":
                return value is null;
            case "IS_EMPTY":
                return ValueTransforms.IsEmptyValue(value);
            case "MATCH":
                return value is not null && !Equals(value, string.Empty) && FindString(value, parameters, processing);
            case "MATCH_REGEX":
                return value is not null && !Equals(value, string.Empty) && FindRegex(value, parameters, processing.Contains("IGNORE_CASE"));
            case "NUMERIC_MATCH":
                return FindNumeric(value, parameters);
            case "NUMERIC_LESS_THAN":
                return Compare(value, parameters, (left, right) => left < right, prefixParse: true);
            case "NUMERIC_MAX":
                return Compare(value, parameters, (left, right) => left <= right, prefixParse: true);
            case "NUMERIC_MORE_THAN":
                return Compare(value, parameters, (left, right) => left > right, prefixParse: true);
            case "NUMERIC_MIN":
                return Compare(value, parameters, (left, right) => left >= right, prefixParse: true);
            case "NUMERIC_RANGE":
                return Range(value, parameters, includeLower: false);
            case "NUMERIC_FULL_RANGE":
                return Range(value, parameters, includeLower: true);
            case "IS_TRUE":
                return Equals(value, 1d) || Equals(value, "1") || Equals(value, true) || Equals(value, "true");
            case "IS_FALSE":
                return Equals(value, 0d) || Equals(value, "0") || Equals(value, false) || Equals(value, "false");
            default:
                return false;
        }
    }

    private static bool FindString(object value, IEnumerable<string> parameters, IReadOnlyCollection<string> processing)
    {
        var text = ValueTransforms.GetValueAsString(value) ?? string.Empty;
        var transformed = (string)ValueTransforms.TransformString(text, processing);
        return parameters.SelectMany(SplitLines)
            .Select(item => (string)ValueTransforms.TransformString(item, processing))
            .Contains(transformed, StringComparer.Ordinal);
    }

    private static bool FindRegex(object value, IEnumerable<string> parameters, bool ignoreCase)
    {
        var text = ValueTransforms.GetValueAsString(value) ?? string.Empty;
        var options = ignoreCase ? RegexOptions.IgnoreCase : RegexOptions.None;
        return parameters.SelectMany(SplitLines).Any(pattern =>
        {
            try
            {
                return Regex.IsMatch(text, pattern, options, RegexTimeout);
            }
            catch (ArgumentException)
            {
                return false;
            }
            catch (RegexMatchTimeoutException)
            {
                return false;
            }
        });
    }

    private static bool FindNumeric(object? value, IEnumerable<string> parameters)
    {
        if (value is null || Equals(value, string.Empty))
        {
            return false;
        }
        var number = ValueTransforms.GetValueAsNumber(value);
        return number is not null && parameters.SelectMany(SplitLines)
            .Select(ValueTransforms.GetValueAsNumber)
            .Any(candidate => candidate == number);
    }

    private static bool Compare(object? value, IReadOnlyList<string> parameters, Func<double, double, bool> compare, bool prefixParse)
    {
        if (value is null || Equals(value, string.Empty) || parameters.Count == 0)
        {
            return false;
        }
        var left = ValueTransforms.GetValueAsNumber(value);
        var right = prefixParse ? ValueTransforms.ParseFloatPrefix(parameters[0]) : ValueTransforms.GetValueAsNumber(parameters[0]);
        return left is not null && right is not null && compare(left.Value, right.Value);
    }

    private static bool Range(object? value, IReadOnlyList<string> parameters, bool includeLower)
    {
        if (value is null || Equals(value, string.Empty) || parameters.Count < 2)
        {
            return false;
        }
        var number = ValueTransforms.GetValueAsNumber(value);
        var lower = ValueTransforms.GetValueAsNumber(parameters[0]);
        var upper = ValueTransforms.GetValueAsNumber(parameters[1]);
        return number is not null && lower is not null && upper is not null &&
               (includeLower ? number >= lower : number > lower) && number <= upper;
    }

    private static IEnumerable<string> SplitLines(string value) => value.Split(["\r\n", "\n"], StringSplitOptions.None);
}
